File: SumORSubset.scala
import java.util.Arrays
object SumORSubset extends App{
    val INT_SIZE = 32
    def getBit(n: Int, k: Int): Int = {
         val bit = if ((n & (1<<k)) !=0)  1 else 0
         bit
     }
    def bitWiseOR(arr: Array[Int]):Int ={
        val bitFreq = new Array [Int] (INT_SIZE)
        val n = arr.length
        var result = 0.0
        for(i <- 0 until n){
            for(k <-INT_SIZE -1 to 0 by -1){
                bitFreq(k) += getBit(arr(i), k)
            }
        }
         for(i <- 0 until INT_SIZE){
           for(count <- bitFreq(i) until 0 by -1){
                result+= Math.pow(2,i)*Math.pow(2,n-count)
            }
         }
         result.toInt

    }
    val arr:Array[Int] = Array(1 , 2 ,3)
    //println(Arrays.toString(arr));
    println(bitWiseOR(arr))
}

File: SumORSubset.java
import java.util.*;
public class SumORSubset{
    static final int BIT_SIZE =32;
    static int getBit(int n, int k){
        return (n & (1<<k)) != 0 ? 1 : 0;
    }
    static int bitWiseOR(  int [] arr ){
        int result = 0;
        int n = arr.length;
        Vector<Integer> bitFreq = new Vector<Integer>(BIT_SIZE);
        for(int i = 0;i<BIT_SIZE;i++){
            bitFreq.add(0);
        } 
        
        for(int i = 0;i<n;i++){
            for (int k = BIT_SIZE-1; k>=0; k--){
                bitFreq.set(k,bitFreq.get(k)+getBit(arr[i],k));
            }
            
         }
        for (int i=0; i < BIT_SIZE; ++i){
            int count = bitFreq.get(i);
            while(count>0){
                result+= Math.pow(2,i)*Math.pow(2,n-count--);
            }
        }
        return result;
    }

public static void main(String args[]){
    int [] arr = { 1, 2, 3};
    System.out.println(bitWiseOR(arr));
    }
}
